File: 2주차/LeetCode_55_JumpGame1/dongjun.java
/**
 * Runtime: 479 ms
 * Memory Usage: 44.4 MB
 * 시간복잡도: nums * nums[] = 최대 30000 * 100000 ?
 *
 *
 * @문제접근
 * dfs로 풀었습니다.
 * 
 */

package ps.leetcode;

public class JumpGame {
    public static void main(String[] args) {
        Solution s = new Solution();
        int[] nums = {8,2,5,1,2,7,5,1,9,9,3,5,0,7,5};

        System.out.println(s.canJump(nums));
    }

}

class Solution {
    public static boolean answer;
    public static boolean visited[];
    public boolean canJump(int[] nums) {
        visited = new boolean[nums.length];
        answer = false;

        dfs(0, nums);

        return answer;
    }

    public static void dfs(int idx, int[] nums) {
        if(idx == nums.length - 1){
            answer = true;
            return;
        }

        for(int i = idx+1; i <= idx+nums[idx]; i++){
            if(i < nums.length && !visited[i]){
                visited[i] = true;
                dfs(i, nums);
            }
        }
    }
}

File: 8주차/LeetCode_78_Subsets/dongjun.py
"""
1. 문제접근
	dfs로 가능한 모든 순열을 만들었습니다.

2. 성능
	Runtime: 28ms
	Memory Usage: 14.2MB
	시간복잡도: O(N!)
"""

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.ret = []
        visited = set()
        
        self.dfs(0, nums, [], visited)
        
        return self.ret
        
    
    def dfs(self, idx, nums, now, visited):
        if len(now) == len(nums):
            self.ret.append(now.copy())
            return
        
        for i, x in enumerate(nums[idx:]):
            if x not in visited:
                now.append(x)
                visited.add(x)
                self.dfs(idx + i, nums, now.copy(), visited)
                del now[-1]
                visited.remove(x)
        
        self.ret.append(now)
